crypto/src/bulletproofs.rs

```diff
@@ -0,0 +1,218 @@
+    }
+}
+
+impl<'de> Deserialize<'de> for PedersenCommitment {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let bytes = Vec::deserialize(deserializer)?;
+        PedersenCommitment::from_bytes(&bytes[..]).map_err(|e| de::Error::custom(e.to_string()))
+    }
+}
+
+impl PartialEq for PedersenCommitment {
+    fn eq(&self, other: &Self) -> bool {
+        self.point == other.point
+    }
+}
+
+impl Eq for PedersenCommitment {}
+
+impl PartialOrd for PedersenCommitment {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.as_ref().partial_cmp(other.as_ref())
+    }
+}
+
+impl Ord for PedersenCommitment {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.as_bytes().cmp(other.as_bytes())
+    }
+}
+
+///
+/// Bulletproof Range Proofs
+///
+
+#[derive(Debug)]
+pub struct BulletproofsRangeProof {
+    proof: RangeProof,
+    bytes: OnceCell<Vec<u8>>,
+}
+
+impl BulletproofsRangeProof {
+    /// Prove that commited value is within the range [0, 2^64). Function only works
+    /// for n = 8, 16, 32, 64
+    pub fn prove_bound(
+        value: u64,
+        blinding: [u8; 32],
+        n: usize,
+        domain: &'static [u8]
+    ) -> Result<(PedersenCommitment, Self), signature::Error> {
+        // Although this is also checked in the bulletproofs library, we check again
+        // to avoid unexpected behaviour in the case of library updates
+        if !(n == 8 || n == 16 || n == 32 || n == 64) {
+            return Err(signature::Error::new());
+        }
+
+        let pc_gens = PedersenGens::default();
+        let bp_gens = BulletproofGens::new(n, 1);
+        let mut prover_transcript = Transcript::new(domain);
+        let blinding = Scalar::from_bits(blinding);
+
+        let (proof, commitment) = RangeProof::prove_single(
+            &bp_gens,
+            &pc_gens,
+            &mut prover_transcript,
+            value,
+            &blinding,
+            n,
+        )
+        .map_err(|_| signature::Error::new())?;
+
+        Ok((
+            PedersenCommitment {
+                point: commitment.decompress().ok_or_else(signature::Error::new)?,
+                bytes: OnceCell::new(),
+            },
+            BulletproofsRangeProof {
+                proof,
+                bytes: OnceCell::new(),
+            },
+        ))
+    }
+
+    // Verify that commited value is within the range [0, 2^64)
+    pub fn verify_bound(
+        &self,
+        commitment: &PedersenCommitment,
+        n: usize,
+        domain: &'static [u8]
+    ) -> Result<(), signature::Error> {
+        if !(n == 8 || n == 16 || n == 32 || n == 64) {
+            return Err(signature::Error::new());
+        }
+
+        let pc_gens = PedersenGens::default();
+        let bp_gens = BulletproofGens::new(n, 1);
+        let mut verifier_transcript = Transcript::new(domain);
+
+        // Create a 64-bit rangeproof.
```

Comment:
does it depend on n? or it's always 64?

Reply:
Yes, this depends on n - I need to start reading the comments I write before submitting PRs. I've been ignoring them and looking at the code in the past.